File: IsomorphicStrings/IsomorphicStrings.js
// Метод доступа size возвращает количество (уникальных) элементов в объекте Set

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function (s, t) {
  let hasht = {};
  for (let i = 0; i < s.length; i++) {
    if (!hasht["s" + s[i]]) hasht["s" + s[i]] = t[i];
    if (!hasht["t" + t[i]]) hasht["t" + t[i]] = s[i];
    if (t[i] != hasht["s" + s[i]] || s[i] != hasht["t" + t[i]]) return false;
  }
  return true;
};


/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function (s, t) {
  let arr1 = {};
  let arr2 = {};
  for (let i = 0; i < s.length; i++) {
    if (arr1[s[i]] != arr2[t[i]]) {
      return false;
    } else {
      arr1[s[i]] = i;
      arr2[t[i]] = i;
    }
  }
  return true;
};


/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function (s, t) {
  let map = new Map();
  if (new Set(s).size != new Set(t).size) return false;
  for (let i = 0; i < s.length; i++) {
    if (map.get(s[i]) && map.get(s[i]) != t[i]) {
      return false;
    } else {
      map.set(s[i], t[i]);
    }
  }
  return true;
};

File: IsomorphicStrings/IsomorphicStrings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(set(s)) != len(set(t)): return False
        hasht = {}
        for i in range(len(t)):
            if t[i] not in hasht:
                hasht[t[i]] = s[i]
            elif hasht[t[i]] != s[i]:
                return False
        return True


class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        dct1, dct2 = {}, {}
        for i, val in enumerate(s):
            dct1[val] = dct1.get(val, []) + [i]
        for i, val in enumerate(t):
            dct2[val] = dct2.get(val, []) + [i]
        return sorted(dct1.values()) == sorted(dct2.values())


class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        return len(set(zip(s, t))) == len(set(s)) == len(set(t))


class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        return [s.find(i) for i in s] == [t.find(j) for j in t]